icu4j/main/tests/core/src/com/ibm/icu/dev/test/util/CodePointTrieTest.java

```diff
@@ -0,0 +1,974 @@
+    }
+
+    /*
+     * Values for testing:
+     * value is set from the previous boundary's limit to before
+     * this boundary's limit
+     *
+     * There must be an entry with limit 0 and the intialValue.
+     * It may be preceded by an entry with negative limit and the errorValue.
+     */
+    private static class CheckRange {
+        CheckRange(int limit, int value) {
+            this.limit = limit;
+            this.value = value;
+        }
+
+        final int limit;
+        final int value;
+    }
+
+    private static int skipSpecialValues(CheckRange checkRanges[]) {
+        int i;
+        for(i=0; i<checkRanges.length && checkRanges[i].limit<=0; ++i) {}
+        return i;
+    }
+
+    private static SpecialValues getSpecialValues(CheckRange checkRanges[]) {
+        int i=0;
+        int initialValue, errorValue;
+        if(i<checkRanges.length && checkRanges[i].limit<0) {
+            errorValue=checkRanges[i++].value;
+        } else {
+            errorValue=0xad;
+        }
+        if(i<checkRanges.length && checkRanges[i].limit==0) {
+            initialValue=checkRanges[i++].value;
+        } else {
+            initialValue=0;
+        }
+        return new SpecialValues(i, initialValue, errorValue);
+    }
+
+    /* ucptrie_enum() callback, modifies a value */
+    private static class TestFilterValue implements CodePointMap.FilterValue {
+        @Override
+        public int apply(int value) {
+            return value ^ 0x5555;
+        }
+    }
+    private static final TestFilterValue testFilter = new TestFilterValue();
+
+    private boolean
+    doCheckRange(String name, String variant,
+                 int start, boolean getRangeResult, CodePointMap.Range range,
+                 int expEnd, int expValue) {
+        if (!getRangeResult) {
+            if (expEnd >= 0) {
+                fail(String.format(  // log_err(
+                        "error: %s getRanges (%s) fails to deliver range [U+%04x..U+%04x].0x%x\n",
+                        name, variant, start, expEnd, expValue));
+            }
+            return false;
+        }
+        if (expEnd < 0) {
+            fail(String.format(
+                    "error: %s getRanges (%s) delivers unexpected range [U+%04x..U+%04x].0x%x\n",
+                    name, variant, range.getStart(), range.getEnd(), range.getValue()));
+            return false;
+        }
+        if (range.getStart() != start || range.getEnd() != expEnd || range.getValue() != expValue) {
+            fail(String.format(
+                    "error: %s getRanges (%s) delivers wrong range [U+%04x..U+%04x].0x%x " +
+                    "instead of [U+%04x..U+%04x].0x%x\n",
+                    name, variant, range.getStart(), range.getEnd(), range.getValue(),
+                    start, expEnd, expValue));
+            return false;
+        }
+        return true;
+    }
+
+    private static int iterStarts[] = { 0, 0xd888, 0xdddd, 0x10000, 0x12345, 0x110000 };
```

Comment:
I like testing all the UTF8 ranges, with before and afters. eg

0, 0x7F, 0x80, 0x7FF, 0x800, 0xFFFF, 0xD7FF, 0xE000, 0x10000, 0x10FFFF, 0x110000

Would suggest adding those to this list.

Reply:
done